exonum/src/runtime/types.rs

```diff
@@ -729,12 +730,176 @@ impl ProtobufConvert for MigrationStatus {
+        instance_id: InstanceId,
+    },
+
+    /// The call is invoked by one of the blockchain lifecycle events.
+    ///
+    /// This kind of authorization is used for `before_transactions` / `after_transactions`
+    /// calls to the service instances, and for initialization of the built-in services.
+    Blockchain,
+
+    // Hidden variant to prevent exhaustive matching.
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+impl Caller {
+    /// Converts a public key to an address.
+    pub fn address_from_key(public_key: PublicKey) -> CallerAddress {
+        Caller::Transaction { author: public_key }.address()
+    }
+
+    /// Returns the author's public key, if it exists.
+    pub fn author(&self) -> Option<PublicKey> {
+        if let Caller::Transaction { author } = self {
+            Some(*author)
+        } else {
+            None
+        }
+    }
+
+    /// Tries to reinterpret the caller as a service.
+    pub fn as_service(&self) -> Option<InstanceId> {
+        if let Caller::Service { instance_id } = self {
+            Some(*instance_id)
+        } else {
+            None
+        }
+    }
+
+    /// Verifies that the caller of this method is a supervisor service.
+    pub fn as_supervisor(&self) -> Option<()> {
+        self.as_service().and_then(|instance_id| {
+            if instance_id == super::SUPERVISOR_INSTANCE_ID {
+                Some(())
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Returns a uniform, forward-compatible presentation of the `Caller` that can be used
+    /// as the account *address*. Different addresses are guaranteed to correspond to
+    /// different `Caller`s.
+    pub fn address(&self) -> CallerAddress {
+        CallerAddress(self.object_hash())
+    }
+}
+
+impl ProtobufConvert for Caller {
+    type ProtoStruct = schema::runtime::Caller;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        let mut pb = Self::ProtoStruct::new();
+        match self {
+            Caller::Transaction { author } => pb.set_transaction_author(author.to_pb()),
+            Caller::Service { instance_id } => pb.set_instance_id(*instance_id),
+            Caller::Blockchain => pb.set_blockchain(Default::default()),
+            Caller::__NonExhaustive => unreachable!("variant is never constructed"),
+        }
+        pb
+    }
+
+    fn from_pb(mut pb: Self::ProtoStruct) -> Result<Self, failure::Error> {
+        Ok(if pb.has_transaction_author() {
+            let author = PublicKey::from_pb(pb.take_transaction_author())?;
+            Caller::Transaction { author }
+        } else if pb.has_instance_id() {
+            Caller::Service {
+                instance_id: pb.get_instance_id(),
+            }
+        } else if pb.has_blockchain() {
+            Caller::Blockchain
+        } else {
+            bail!("No variant specified for `Caller`");
+        })
+    }
+}
+
+/// Uniform presentation of a `Caller`.
```

Comment:
It will be nice to have an explanation how clients (e.g. light clients) can obtain `CallerAddress, having the `PublicKey` of recipient.